src/libstd/io/buffered.rs

```diff
@@ -818,6 +894,261 @@ impl<W> fmt::Display for IntoInnerError<W> {
     }
 }
 
+/// Private helper struct for implementing the line-buffered writing logic.
+/// This shim temporarily wraps a BufWriter, and uses its internals to
+/// implement a line-buffered writer (specifically by using the internal
+/// methods like write_to_buf and flush_buf). In this way, a more
+/// efficient abstraction can be created than one that only had access to
+/// `write` and `flush`, without needlessly duplicating a lot of the
+/// implementation details of BufWriter. This also allows existing
+/// `BufWriters` to be temporarily given line-buffering logic; this is what
+/// enables Stdout to be alternately in line-buffered or block-buffered mode.
+#[derive(Debug)]
+pub(super) struct LineWriterShim<'a, W: Write> {
+    buffer: &'a mut BufWriter<W>,
+}
+
+impl<'a, W: Write> LineWriterShim<'a, W> {
+    pub fn new(buffer: &'a mut BufWriter<W>) -> Self {
+        Self { buffer }
+    }
+
+    /// Get a reference to the inner writer (that is, the writer wrapped by
+    /// the BufWriter)
+    fn inner(&self) -> &W {
+        self.buffer.get_ref()
+    }
+
+    /// Get a mutable reference to the inner writer (that is, the writer
+    /// wrapped by the BufWriter). Be careful with this writer, as writes to
+    /// it will bypass the buffer.
+    fn inner_mut(&mut self) -> &mut W {
+        self.buffer.get_mut()
+    }
+
+    /// Get the content currently buffered in self.buffer
+    fn buffered(&self) -> &[u8] {
+        self.buffer.buffer()
+    }
+}
+
+impl<'a, W: Write> Write for LineWriterShim<'a, W> {
+    /// Write some data into this BufReader with line buffering. This means
+    /// that, if any newlines are present in the data, the data up to the last
+    /// newline is sent directly to the underlying writer, and data after it
+    /// is buffered. Returns the number of bytes written.
+    ///
+    /// This function operates on a "best effort basis"; in keeping with the
+    /// convention of `Write::write`, it makes at most one attempt to write
+    /// new data to the underlying writer. If that write only reports a partial
+    /// success, the remaining data will be buffered.
+    ///
+    /// Because this function attempts to send completed lines to the underlying
+    /// writer, it will also flush the existing buffer if it contains any
+    /// newlines, even if the incoming data does not contain any newlines.
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let newline_idx = match memchr::memrchr(b'\n', buf) {
+            // If there are no new newlines (that is, if this write is less than
+            // one line), just do a regular buffered write
+            None => {
+                // Check for prior partial line writes that need to be retried.
+                // Only retry if the buffer contains a completed line, to
+                // avoid flushing partial lines.
+                if let Some(b'\n') = self.buffered().last().copied() {
+                    self.buffer.flush_buf()?;
```

Comment:
This only checks the last character in the buffer. Isn't it possible for a previous line write attempt to fail and a partial line being at the end of the buffer, e.g. because the buffer got filled when the `write_to_buf` call at the end of this method attempted to write two lines and only the first line + some part of the second fit into the buffer?

Reply:
Hmm. That's a good point; I only write out to the last newline at the end of this method, but I didn't consider the case that the `[flushed..newline_idx]` range might be larger than the buffer.

Reply:
Although, in that case, the buffer is going to get unconditionally flushed anyway, because it's full, right?

Reply:
Ah true. But that could lead to partial lines being written on flush. So it might be better to only try to buffer either less than one line up to as many whole lines as fit into the buffer. That way lines would not get torn unless they exceed the buffer size (or the atomic write size of the underlying IO primitive).

Additionally if the underlying writer supports vectored writes it might also be good to try to write the remainder of the buffer + the new write in a single vectored request to mend torn lines.

Reply:
Yeah, I see what you mean. I went back and forth a lot with how to handle the different ways that partial line writes can occur- and luckily, the `LineWriter` design I'm replacing is wildly inconsistent about that, so I didn't feel bad making changes. One of the conclusions that I arrived at (obvious in retrospect) is that at the very least you have to allow partial line writes if a single line is longer than the buffer.

But I see what you mean that *mutliple* lines could be added to the buffer, filling it, but the last character isn't a newline, so now we're invoking the "write a partial line" logic I just described to several lines. I'll see about adjusting the logic to "buffer complete lines out to the last one that fits in the buffer".

> Additionally if the underlying writer supports vectored writes it might also be good to try to write the remainder of the buffer + the new write in a single vectored request to mend torn lines.

Yeah, I noticed this too. I actually think it's out of scope for this particular pull request, but I have another branch in progress where I add this behavior to `BufWriter` and want to add it to `LineWriter` too. It ends up being a bit complex- not so much that it's not worth doing, but it's more than the trivial conditional at the top of `write` that I thought it'd be.

Reply:
Fixed, with tests!